### PR TITLE
Fix rotation when auto-rotate is disabled

### PR DESCRIPTION
**Pull request by blackbox87**
#### What is it?
- [x] Bug fix (user facing)
- [ ] Feature (user facing)
- [ ] Code base improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR

1. Enable auto-rotate and watch a video in landscape
2. Disable auto-rotate while watching the video in landscape

Previously NewPipe would return you to your default orientation, but now you'll remain stuck in landscape and NewPipe's fullscreen button won't do anything no matter how many times you press on it.

With this change NewPipe will remain in landscape mode, but when you press on the fullscreen button it'll correctly change the orientation to portrait.

#### Testing apk
[app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5196379/app-debug.zip)

#### Agreement
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

### CODE REVIEW

**Comment by opusforlife2** [comment]
APK to test.

**Comment by avently** [comment]
Works, thanks!

**Comment by opusforlife2** [comment]
Yup! Thanks for the fix!

**Comment by opusforlife2** [comment]
Oh no. Bug. First two steps are the same as yours.
1. Enable auto-rotate and watch a video in landscape
2. Disable auto-rotate while watching the video in landscape

Then tap back. The app will try to rotate to portrait, but then fail and go back to landscape. It will continue playing through this, and if you pause and tap back, it will start playing when it fails and goes back to landscape.

In the latest unified debug apk, tapping back in this situation pauses the video and shows you video details in landscape.

**Comment by avently** [comment]
@opusforlife2 nice catch! I would really like to see you using my app. Do you like trading on crypto?:)

**Comment by opusforlife2** [comment]
@avently For that you need money of your own first. ( ;´༎ຶД༎ຶ`)

Which app, though?

**Comment by blackbox87** [comment]
> Then tap back. The app will try to rotate to portrait, but then fail and go back to landscape. It will continue playing through this, and if you pause and tap back, it will start playing when it fails and goes back to landscape.

That seems to be an existing issue that's caused by the fullscreen/rotate button and the back button behaving differently. I'll see if I can fix that too though.

**Comment by opusforlife2** [comment]
If you mean that the back button pauses the video, but the full screen button keeps playing it when you go back to portrait, I think that's intentional. Am I wrong, @avently?

**Comment by blackbox87** [comment]
> If you mean that the back button pauses the video, but the full screen button keeps playing it when you go back to portrait, I think that's intentional.

Yes. But I'm not sure what else you'd mean?

If I do steps 1 & 2 and then press my back button it'll correctly change to portrait. And then if I want to go back to landscape all I need to do is tap on the fullscreen button again. That works perfectly for me on my device and emulators.

**Comment by opusforlife2** [comment]
Oh. I'll make a screenrecord, then.

**Comment by blackbox87** [comment]
@opusforlife2 This is what it looks like on my device and emulators.

rotation_fix

I can only get it to do what you describe if I launch the unpatched dev version of NewPipe.

**Comment by opusforlife2** [comment]
[ScreenRecord.mp4.gz](https://github.com/TeamNewPipe/NewPipe/files/5197530/ScreenRecord.mp4.gz)

Here's my screen record. I have shown the problem happening twice.

**Comment by blackbox87** [comment]
@opusforlife2 What phone is that and which version of Android is it using?

At the moment I can only assume that it's a device specific issue. It's as if applying SCREEN_ORIENTATION_PORTRAIT and then immanently reverting back to a user or sensor preference.

**Comment by opusforlife2** [comment]
Xiaomi Mi 5 running LOS 16.

**Comment by avently** [comment]
> For that you need money of your own first. ( ;´༎ຶД༎ຶ`)

There is a demo mode included (in-app exchange), even zero balance on real exchange is ok for start learning:)

>Which app, though?

Called `Evraon`. In Google Play already, so you can find it if you want.

**Comment by blackbox87** [comment]
@opusforlife2 Try this. It looks like the issue was caused by some existing code, which shouldn't be required.

[app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5220744/app-debug.zip)

**Comment by opusforlife2** [comment]
Worked! Thanks!

**Comment by opusforlife2** [comment]
So what was this code conflicting with?

**Comment by blackbox87** [comment]
@opusforlife2 Just pushed the changes. But the conflicting code was this...

```java
        // This will show systemUI and pause the player.
        // User can tap on Play button and video will be in fullscreen mode again
        // Note for tablet: trying to avoid orientation changes since it's not easy
        // to physically rotate the tablet every time
        if (!DeviceUtils.isTablet(activity)) {
            activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED);
        }
```

It's unnecessary because toggleFullscreen() sets the orientation under the correct conditions and this chunk of code would undo that.

**Comment by opusforlife2** [comment]
Ah. Yes. The simple and straightforward orientation switch button was replaced with the fullscreen button in the unified player, which contains additional logic to detect orientation.

**Comment by blackbox87** [comment]
> Ah. Yes. The simple and straightforward orientation switch button was replaced with the fullscreen button in the unified player, which contains additional logic to detect orientation.

Yeah. Personally I prefer having separate options like in v0.19.8 as it gives you more control and it seems less buggy, but at least this commit fixes some of the pitfalls.

Just try to watch a vertical video in landscape with auto-rotate disabled. You can't do it anymore and that seems to be by design..

**Comment by opusforlife2** [comment]
Well, yeah? Why would you want to do that? Unless you're trying to practise squinting to exercise your eye muscles. :P

**Comment by blackbox87** [comment]
With v0.19.8 if you had a playlist that contained mostly landscape videos and a few portrait ones then they'd all be played in landscape, so it was consistent and honored your preferred orientation until you stopped playing videos. If you try to do that now though then the second a portrait video is played you'll be taken out of landscape/fullscreen so that the orientation can be corrected.

I like things to work in a consistent way across all devices and this new unified player just can't do that anymore.

**Comment by opusforlife2** [comment]
Holy poop that is buggy! I just tried it. When you have auto-rotate locked to portrait and you're watching a landscape video in fullscreen, if the next video is a portrait one, the video stretches to fill the screen, then the player goes back to the video details fragment. I was expecting it to rotate to portrait but stay in fullscreen.

But everything works bug free if auto-rotate is on. Portrait videos play normally in fullscreen landscape. And landscape videos play normally in fullscreen portait. No forced rotation occurs.

You just gave me a new avenue of testing the app by breaking it. Thanks. 🤣

**Comment by blackbox87** [comment]
Yeah, it's just really janky. I don't know what can be done about it though, unless it was reverted to how it's done in v0.19.8.

This commit is ready for merging as it puts the unified player in a slightly better state than it was, but if anyone else can come up with a better solution then I'd close the PR.

**Review by Stypox — APPROVED** [review]
Thank you! The apk works fine on my 7.0 phone, on 10 phone and tablet emulator, and code looks good :-D
@avently could you confirm this can be merged?

**Comment by Stypox** [comment]
Actually, I have found an issue, probably unrelated to this PR: the fullscreen button is missing in tablets in portrait, thus requiring to rotate the phone to go to landscape (which for tablets does not mean fullscreen) and then press fullscreen. I don't know if this is intended behaviour, since it does make some sense. @avently @blackbox87 what do you think?

**Comment by blackbox87** [comment]
@avently Yea, it's unrelated to this PR. I noticed it myself, which is one of the reasons why I said that rotation behaviour is "janky" at the moment.

https://github.com/TeamNewPipe/NewPipe/blob/e0f02d40805f121180a6ef75e24421989386b00f/app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java#L926-L929

The code was added by @avently.

**Comment by blackbox87** [comment]
@Stypox Yea, it's unrelated to this PR. I noticed it myself, which is one of the reasons why I said that rotation behaviour is "janky" at the moment.

https://github.com/TeamNewPipe/NewPipe/blob/e0f02d40805f121180a6ef75e24421989386b00f/app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java#L926-L929

The code was added by @avently.

**Comment by avently** [comment]
@Stypox 
>  have found an issue, probably unrelated to this PR

Unrelated to this PR and @blackbox87 can't do anything with this (at least not with a correct way) until #4288 is merged. Because here after screen orientation change the player is null https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java#L1496

After performance PR is merged in this place the player will not be null after fragment creation (for the second time) so you'll be able to check fullscreen state correctly.

> could you confirm this can be merged?

Unfortunately, no, this can't be merged in it's current state:
- @blackbox87 you added some custom behavior into `toggleFullscreen()` which is used many times in the app and the app doesn't expect such custom things. I prefer that such important methods should do as small piece of work as possible
- right now if you have locked orientation in landscape, and watching a video in landscape (fullscreen), and then long pressing on `arrow down` button (from top right corner), then pressing on background playback or popup button, you'll see some weird orientation changes
- since you probably need to rotate the orientation from landscape locked to portrait locked only in one place you can do it without code inside `toggleFullscreen()` but with adding the code to, for example, https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java#L791 (and in other places if needed)
- `void restoreDefaultOrientation` should be renamed to something like `exitFromFullscreenIfNeeded`

**Comment by blackbox87** [comment]
@avently Maybe you could try to do it yourself from #4288 since that includes orientation changes already?

I'm not familiar enough with the code, which is why it was done this way in the first place. And I don't agree with your vision for how any of this should behave, so whatever I do is probably going to conflict with how you want it done.

**Comment by avently** [comment]
@blackbox87 it looks like you always blame my code but unable to do better, am I right? Here and there I see that you disagree with me but can't give a better alternative. I'm ok with that and understand your intent - you want to create something great too, so it's fine. But take a look at the code and you can find a better way than already exists, I believe you can.

> Maybe you could try to do it yourself

I can try to do everything but it's not what I want to do.

>  since that includes orientation changes already

It doesn't. It's about performance, not orientation.

> I'm not familiar enough with the code

I wasn't familiar with the code when I started Unified player some years ago too.

> so whatever I do is probably going to conflict with how you want it done

It's not true based on how I feel like about your code. It's pretty fine, no conflicts. You just missed some important context and I said about that. No problem here.

**Comment by blackbox87** [comment]
> it looks like you always blame my code but unable to do better, am I right?

> Here and there I see that you disagree with me but can't give a better alternative

I suggest alternatives, but I'm not about to put time into making pull requests while you're making pull requests that contain multiple fixes instead of seperate pull requests for each fix. It creates too much conflict.

I don't like the changes that you've made to the rotation behaviour, but your changes keep getting merged and I don't have the time to go back and forth over issues. So this PR can either be accepted as it is (since it does fix the bug described) or closed.

**Comment by avently** [comment]
> since it does fix the bug described

But the fact that it makes some more bugs it's ok for you, right?

**Comment by avently** [comment]
> but I'm not about to put time into making pull requests while you're making pull requests that contain multiple fixes instead of separate pull requests for each fix. It creates too much conflict.

Your PR will be merged before my PRs so it's not an argument here. I will base my PRs on top of your changes.

**Comment by blackbox87** [comment]
> But the fact that it makes some more bugs it's ok for you, right?

I'm fine with it, since #4154 was knowingly merged with issues. And you say that further changes will be required to fix the issue that @Stypox mentioned and the [one I found](https://github.com/TeamNewPipe/NewPipe/pull/4288#issuecomment-695113779), so someone else can do that once #4288 is also merged.

**Comment by avently** [comment]
> I'm fine with it, since #4154 was knowingly merged with issues

It's not merged with issues, it didn't fix ALL issues.

> And you say that further changes will be required to fix the issue

I said that it's not related to your PR and you shouldn't do anything with it. 

And I already fixed the issue mentioned by @Stypox in my local branch on top of performance PR. I also fixed #4322. I just thinking about to upload the changes inside performance PR or to wait until it gets merged because it's unrelated to performance.

**Comment by blackbox87** [comment]
> It's not merged with issues, it didn't fix ALL issues.

I posted about the issues with it before it was merged, which is why I say that it was knowingly merged with issues.

Like if I go back to you saying this...
> right now if you have locked orientation in landscape, and watching a video in landscape (fullscreen), and then long pressing on `arrow down` button (from top right corner), then pressing on background playback or popup button, you'll see some weird orientation changes

I don't see any weird orientation changes, which is why I suggested that you should fix it yourself in #4288 since that already includes orientation changes.

I don't want to keep making changes to this PR and would honestly rather close it, since [like I've already explained](https://github.com/TeamNewPipe/NewPipe/pull/4288#issuecomment-695048750), I don't like how rotation is handled anymore. And at this point I have no desire to contribute anything more to the project and would rather help out with NewPipe Legacy where possible, since that might not support the unified player due to API requirements.

**Comment by avently** [comment]
@Stypox regarding the fullscreen button in tablets: [perf-with-orientation.zip](https://github.com/TeamNewPipe/NewPipe/files/5287514/perf-with-orientation.zip)

This apk is built on top of performance PR. It will work on top of #4272 partially. So I can merge this code into #4272 but it will work fully only after merge of #4288. I don't have a time on waiting the merge of Performance PR so I think it's ok, what do you think?

**Comment by avently** [comment]
> I don't like how rotation is handled anymore

What's you vision? How you would do all rotation logic?

**Comment by blackbox87** [comment]
> What's you vision? How you would do all rotation logic?

https://github.com/TeamNewPipe/NewPipe/pull/4288#issuecomment-695048750

I want it to behave exactly as it did in v0.19.8. Even if that means that when auto-rotate is disabled you'll see an additional button.

As things are I'd rather use NewPipe Legacy or fork NewPipe and revert all of the unified player changes.

**Comment by avently** [comment]
@blackbox87  I fixed the issue that you tried to fix in this PR in a couple of lines of code. Do you want me to apply my changes to Small fixes2 or you'll want to find a solution yourself? You did correct things but in incorrect place.

> As things are I'd rather use NewPipe Legacy or fork NewPipe and revert all of the unified player changes.

It's sad to hear that you don't like the unified player. I tried to make it awesome and useful. I believe that you'll find it useful too.

**Comment by blackbox87** [comment]
> @blackbox87 I fixed the issue that you tried to fix in this PR in a couple of lines of code. Do you want me to apply my changes to Small fixes2 or you'll want to find a solution yourself? You did correct things but in incorrect place.

You can apply it to your PR and I'll close this one. That way everything is where you'd like it to be and it's easier for you to maintain.

> It's sad to hear that you don't like the unified player. I tried to make it awesome and useful. I believe that you'll find it useful too.

As someone who often has auto-rotate disabled I just really don't like how the merged rotate/fullscreen button changes depending on the type of device you're using or if you're playing a vertical or horizontal video. It's very inconsistent and seems unintuitive to me.

It'll be interesting to see if others agree with me when an update goes out to everyone.

> If I started a playlist, locked my device to landscape and then put my device down then I'd expect every video to honor my orientation choice by playing in landscape. If it doesn't do that then it might eventually get to a vertical video and suddenly the image on my screen is sideways, which requires me to alter how I'm holding the device. So that doesn't respect my orientation choice and it's going to annoy people.
> 
> Use NewPipe v0.19.8 with your orientation locked to portrait, start a mixed playlist and then press on NewPipe's button to change the orientation to landscape. It'll then respect your orientation choice and play every video in landscape until you exit.

**Comment by avently** [comment]
@blackbox87 
Uploaded changes: 
https://github.com/TeamNewPipe/NewPipe/pull/4272#issuecomment-699570617

@Stypox take a look too. Fixed the issue you mentioned.

Fix in one line: https://github.com/TeamNewPipe/NewPipe/pull/4272/commits/df9823988ef3c386a5c4d3c4b18946937ba42276#diff-57eb59a31e38fed8f965fb4a544e3b9fL1965

**Comment by blackbox87** [comment]
@avently It fixes what's described in my original post, but it doesn't fix the secondary issue.

- Enable auto-rotate and play a vertical video
- Rotate your device so that it's landscape and then disable auto-rotate
- Press the rotate/fullscreen button

You'd expect it to rotate the fullscreen video since that's what horizontal videos always do, but instead it remains acting as if you're wanting to use fullscreen.

https://github.com/TeamNewPipe/NewPipe/pull/4254/files#diff-61ebd1c28e394d4dbe7ffd9c9402e610

That's why I put the fix where it is. It catches the issue and makes the button rotate the video instead.

**Comment by avently** [comment]
> You'd expect it to rotate the fullscreen video since that's what horizontal videos always do, but instead it keeps taking you in and out of fullscreen.

I don't expect but if you do here is what should be done in order to support this use-case:

Replace this line https://github.com/TeamNewPipe/NewPipe/blob/df9823988ef3c386a5c4d3c4b18946937ba42276/app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java#L809
with:
```java
if (!isVerticalVideo || (!DeviceUtils.isTablet(service) && service.isLandscape())) {
```

If this fixes the problem you mentioned and it will make you happy I'll make a new commit in smallFixes2. Let me know.

Don't expect that the vertical video will stay in fullscreen after rotation because it is fixed inside performance PR (not in smallFixes2).

**Comment by blackbox87** [comment]
> I don't expect but if you do here is what should be done in order to support this use-case

If you disabled auto-rotate while watching a horizontal video then it'll rotate, but if you do the same thing while watching a vertical video in landscape then you'll be stuck in landscape and the button doesn't. So this restores some consistency.

The lack of consistency with the unified player is why I said that it seems unintuitive to me earlier.

> If this fixes the problem you mentioned and it will make you happy I'll make a new commit in smallFixes2. Let me know.

I did a quick test with the suggested change on my device and it seems to work.

**Comment by avently** [comment]
> So this restores some consistency

Agree, Added the commit https://github.com/TeamNewPipe/NewPipe/pull/4272/commits/9a0c2c40bd6fa6e5f77fd9ee317ba8ebb881aba8

Final code is a little bit different, this line https://github.com/TeamNewPipe/NewPipe/pull/4272/commits/9a0c2c40bd6fa6e5f77fd9ee317ba8ebb881aba8#diff-61ebd1c28e394d4dbe7ffd9c9402e610R818

**Comment by blackbox87** [comment]
@avently I can't seem to trick it to get stuck in landscape, so it's working perfectly. Thanks.

I'll close this PR and hopefully they can get #4272, #4288 and #3178 merged soon.

**Comment by TobiGr** [comment]
@blackbox87 thanks for the effort!

**Pull request closed without merging**
